### PR TITLE
Add JWT bearer authorization

### PR DESCRIPTION
**Pull request by kthomas**
This effort adds the ability for a third-party to sign JWT tokens containing well-formed `permissions` within the payload and run NATS server configured with the public key for signature verification.

The initial version of this PR represents the result of a quick and dirty spike to get something that could be tested against a fast-moving project in the blockchain space. It makes sense to consider this functionality because it allows applications to leverage the existing NATS permissions model without the need for an account server. This is an alternative strategy to nkey auth.

When "bearer authorization mode" is configured, a connection to NATS will fail to be authorized under the following conditions:
- when no bearer token is provided by the client
- when a bearer token is provided but fails signature verification based on the configured `JWT_SIGNER_PUBLIC_KEY`
- when the bearer token is verified but no `permissions` claim exists in the JWT

The authorization model under this strategy is delegated to NATS by the application, and JWT tokens which an application vends need to be issued with a short ttl; applications and NATS clients need to consider the refresh cycle. I am considering how to best support an interface for defining a configurable callback within NATS clients which will ask the application to vend a new token (i.e. "refresh" the soon-to-expire token) at the appropriate time and gracefully recycle the socket to establish a new NATS connection using the newly-authorized token, just prior to the previously-valid token `exp` timestamp.

There are some obvious things that need to be resolved in the course of diligence here that I am working on:
- <s>removing the hack of adding a third-party JWT package directly instead of updating the NATS JWT package</s>
- <s>supporting JetStream</s>
- <s>providing broad client support for gracefully refreshing bearer tokens just before they expire as described above</s>

If you want to run the branch to try it out, you can check it out, `go build` and then run the following:
```
➜  nats-server git:(jwt-bearer-auth) JWT_SIGNER_PUBLIC_KEY='
-----BEGIN PUBLIC KEY-----
MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAqU/GXp8MqmugQyRk5FUF
BvlJt1/h7L3Crzlzejz/OxriZdq/lBNQW9S1kzGc7qjXprZ1Kg3zP6irr6wmvP0W
YBGltWs2cWUAmxh0PSxuKdT/OyL9w+rjKLh4yo3ex6DX3Ij0iP01Ej2POe5WrPDS
8j6LT0s4HZ1FprL5h7RUQWV3cO4pF+1kl6HlBpNzEQzocW9ig4DNdSeUENARHWoC
ixE1gFYo9RXm7acqgqCk3ihdJRIbO4e/m1aZq2mvAFK+yHTIWBL0p5PF0Fe8zcWd
NeEATYB+eRdNJ3jjS8447YrcbQcBQmhFjk8hbCnc3Rv3HvAapk8xDFhImdVF1ffD
FwIDAQAB
-----END PUBLIC KEY-----' ./nats-server -DV
```

 - [x] Link to issue: related to https://github.com/nats-io/nats-server/issues/434
 - [ ] Documentation added (in progress)
 - [x] Tests added
 - [x] Changes squashed to a single commit (described [here](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))
 - [x] Build is green in Travis CI
 - [x] You have certified that the contribution is your original work and that you license the work to the project under the [Apache 2 license](https://github.com/nats-io/gnatsd/blob/master/LICENSE)

/cc @nats-io/core

### CODE REVIEW

**Comment by derekcollison** [comment]
Will take a look today, thanks for the patience.

**Comment by kthomas** [comment]
No rush. I still need to add some tests etc.

**Comment by kozlovic** [comment]
@kthomas I had a look, but this is early stage WIP, so no much to comment on. I would pass the server or logger to the BearerAuthFactory() function so that you can then use the logger instead of fmt.Println() calls.

**Comment by kthomas** [comment]
Thanks @kozlovic -- I had begun poking around looking for how to best use the package logger. Makes sense to pass it along.

I will be pushing another commit or two shortly to clean things up, including tests, and we can discuss.

**Comment by wuhenfeike** [comment]
I see that the current sub and push limit are set when the client login, can it be changed to set the sub and push permission at any time？

**Comment by wuhenfeike** [comment]
Or when subscribing or pushing to the server, the server can query the user's push and subscription rights from mysql or other databases to restrict the push and subscription rights。

**Comment by wuhenfeike** [comment]
@kthomas Or when subscribing or pushing to the server, the server can query the user's push and subscription rights from mysql or other databases to restrict the push and subscription rights。

**Comment by kthomas** [comment]
@wuhenfeike if I understand your question correctly, it sounds like it could be supplied within your bearer token similarly to how `permissions` is being presented. You could issue a new token to update it I presume.

That is out of scope for the PR as of this moment but I will take a look to see how it could be supported as I groom this further. It seems you could present those limits as part of the signed token payload as well.

I didn't follow the question related to re-centralizing on a SQL database, you lost me there.

**Comment by wuhenfeike** [comment]
Thanks @kthomas For example, in iot scenario application, two clients can know the restricted objects of push and subscription after determining the binding relationship, so whether can provide the function of dynamically setting push and subscription limits.

**Comment by kozlovic** [comment]
@kthomas Just a note: although we have go.mod, as of now, we build/test from vendor directory. So you would have to `go mod vendor` and add the jwt into vendor in the commit if you want Travis to run the tests...

**Comment by kthomas** [comment]
Have some tests to commit as well @kozlovic -- ty for this.

**Comment by kthomas** [comment]
@derekcollison @kozlovic this is probably ready for another pass when you have a chance.

**Comment by hastarin** [comment]
Excuse my ignorance but is this something that could tie in with an external service like auth0.com to better fit NATS server into a OIDC/OAuth 2.0 flow?

**Comment by kthomas** [comment]
@hastarin this is meant to be very primitive for portability's sake.

The intent of this PR is indeed to allow third-party services to generate and sign JWT credentials (in essence delegating the NATS authorization to some service without having to run extra NATS account server infrastructure). So if the third-party service supports generating JWT tokens with arbitrary claims and signing them using the `RS256` algo, then such a service would indeed be supported.

Hope this answers your question.

File: server/bearer_auth.go [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+	bearer.server.Debugf(fmt.Sprintf("parsed bearer authorization: %s\n; client authentication: %s", jwtToken.Claims, c))
+	claims, claimsOk := jwtToken.Claims.(jwt.MapClaims)
+	if !claimsOk {
+		bearer.server.Warnf(fmt.Sprintf("no claims present in verified JWT; %s", err.Error()))
+		return false
+	}
+
+	permissions := &Permissions{}
+	if permissionsClaim, permissionsClaimOk := claims["permissions"].(map[string]interface{}); permissionsClaimOk {
```
Comment (kthomas) [inline server/bearer_auth.go]: FIXME... the `permissions` claim should be tucked underneath the `nats` claim.

i.e., here is a current JWT payload as generated by `nsc`:
```
{
  "jti": "7WA3QTYWKVEEETMNBJENR34L5SQQYYCE7QB5H7MLEIDBYHX5XGUA",
  "iat": 1576587866,
  "iss": "OCZM6CMZMDLBEECLABZZOLKF7RFTUOWWEL3BHBOKL6GS6VMM56V6X3BB",
  "name": "A",
  "sub": "ABPAHQBU2XAHCENGXGV7NITB4PR2CNEMO3RZKTYJXA3WUR3Q56Y7QUEU",
  "type": "account",
  "nats": {
    "limits": {
      "subs": -1,
      "conn": -1,
      "leaf": -1,
      "imports": -1,
      "exports": -1,
      "data": -1,
      "payload": -1,
      "wildcards": true
    }
  }
}
```

I propose that under this bearer permissions strategy, the `nats` claim encoded in vended JWTs should look like this (other standard JWT claims omitted for brevity):
```
  "nats": {
    "limits": {
      "subs": -1,
      "conn": -1,
      "leaf": -1,
      "imports": -1,
      "exports": -1,
      "data": -1,
      "payload": -1,
      "wildcards": true
    },
    "permissions": {
      "publish": {
        "allow": [],
        "deny": []
      },
      "subscribe": {
        "allow": [],
        "deny": []
      },
      "responses": {
        "max": 1,
        "ttl": 120000000000
      },
    }
  }
}
```

**Comment by kthomas** [comment]
Here is some additional context on this effort after several discussions with the NATS team:

**Goals:**

- to lay the architectural groundwork to facilitate increasingly distributed NATS ecosystems

- to first enable a single, ephemeral global account ([like this](https://github.com/nats-io/nats-server/pull/1149/commits/8ac7467644ffaa55420c3058acd31f50678be2ff#diff-ed427ec66d9bd8da6e989d0aef90abccR87-R90)) by verifying bearer JWTs signed using the RS256 algorithm (architecturally, this could support any JWT signing algorithm i.e., via the `alg` and `kid` headers)

- to support this bearer authorization pattern (with a signed `nats` [permissions claim](https://github.com/nats-io/nats-server/pull/1149/commits/8ac7467644ffaa55420c3058acd31f50678be2ff#r360669657))

- to support the above strategy in a standalone (_"configurationless"_) mode as well as with NATS nkey-based bearer tokens (i.e., nkey-based JWTs which are vended without a challenge–response)

- to add support for the above bearer authorization strategies to simultaneously support NATS websockets

**Challenges:**

- nkey is a "proprietary" format and supports only a single signing alg, and that alg is not RS256

- the nkey format couples the concept of NATS identity to the token vending mechanism

- the addition of "raw" JWT support (meaning, decoupled from NATS identities) will add overhead to LTS of the NATS project if not implemented thoughtfully

- the decision to support additional JWT signing algorithms has implications to security within the NATS ecosystem

This PR is proposing we add only initial support for RS256 and to allow that "mode" of operation to function alongside new nkey bearer functionality.

^^ thanks @derekcollison @aricart @kozlovic @ColinSullivan1

**Comment by ebekker** [comment]
Was curious what the state of this work is?  It looks like there was a bit of activity for a while and several other features as well that had dependencies on this, but all have gone quiet for a while now.

**Comment by aricart** [comment]
> Was curious what the state of this work is? It looks like there was a bit of activity for a while and several other features as well that had dependencies on this, but all have gone quiet for a while now.

We are implementing Bearer tokens for now, not part of Kyles fork.

**Comment by kthomas** [comment]
> Was curious what the state of this work is? It looks like there was a bit of activity for a while and several other features as well that had dependencies on this, but all have gone quiet for a while now.

There is a fork with this effort [on this branch](https://github.com/kthomas/nats-server/tree/bearer-auth). There is also a docker image you can try [here](https://hub.docker.com/r/provide/nats-server). The fork is still being maintained at this time while the feature finds its way into NATS-proper. Hope this is helpful!

**Comment by kthomas** [comment]
@derekcollison @aricart would it be helpful for this PR to be updated?

**Comment by derekcollison** [comment]
Possibly.

**Comment by ebekker** [comment]
Looking forward to when this feature will become officially released.  In the meantime I had a short-term need to support authentication based on Kerberos and AWS IAM credentials.  So I put together this [stopgap solution](https://github.com/ebekker/NATS.Mapper) based on the existing NKeys support, in case anyone is interested.

**Comment by kthomas** [comment]
> Looking forward to when this feature will become officially released. In the meantime I had a short-term need to support authentication based on Kerberos and AWS IAM credentials. So I put together this [stopgap solution](https://github.com/ebekker/NATS.Mapper) based on the existing NKeys support, in case anyone is interested.
^^ that is pretty interesting workaround @ebekker... thanks for sharing 👍 

@kozlovic can we explore what it would take to merge this PR? The fork has +10K pulls on dockerhub...
image

I would be happy to help get it over the finish line if there is any feedback, although I think you will find there is sufficient rigor in the tests.

This PR has been open for nearly a whole year...  🎉 🎂

**Comment by kozlovic** [comment]
Not sure why this is still opened because I believe we said that we are not going this direction in the past. I will let @derekcollison have the final word.

**Comment by kthomas** [comment]
> Not sure why this is still opened because I believe we said that we are not going this direction in the past. I will let @derekcollison have the final word.

Agree, but since it's about to celebrate it's first birthday maybe we take that as a sign and discuss if it has a place at this time...

I am good either way but it would help a whole lot if we could agree to merge it ;)

Thanks, as always 🙏

**Comment by kthomas** [comment]
@derekcollison @kozlovic

This has been rebased on latest master. Will fix failing tests.

**Comment by kthomas** [comment]
image

Can't compete with 50M+ pulls on the official repo :)

We should compare notes related to this PR once and for good, @kozlovic.

cc: @derekcollison

**Comment by kozlovic** [comment]
@kthomas Again, I think that this is a discussion that you have to have with @derekcollison. I let him know and he will touch base with you and maybe pull some other resources into a meeting.

**Comment by snehesht** [comment]
Any thoughts on the timeline for this PR ?

**Comment by derekcollison** [comment]
As our focus shifts to the marketplaces etc this will come back around. That should be in next few weeks.

**Pull request closed without merging**